File: tests/omp/critical2.cpp
#include <omp.h>

int dequeue(float *a);
void work(int i, float *a);

void critical_example(float *x, float *y)
{
  int ix_next, iy_next;

  {
    ix_next = dequeue(x);
    work(ix_next, x);

    iy_next = dequeue(y);
    work(iy_next, y);
  }

}

File: tests/omp/critical2.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->
  
  !! Omp.parallel [Shared ["x";"y"]; Private ["ix_next";"iy_next"]] [tAfter; cVarDef "ix_next"];
  !! Omp.critical "xaxis" ~hint:"omp_lock_hint_contented" [tBefore; sInstr "ix_next = dequeue(x)"];
  !! Omp.critical "yaxis" ~hint:"omp_lock_hint_contented" [tBefore; sInstr "iy_next = dequeue(y)"];
)
